src/udev/udev-builtin-net_id.c

```diff
@@ -305,54 +421,10 @@ static int dev_pci_slot(sd_device *dev, struct netnames *names) {
         if (l == 0)
                 names->pci_path[0] = '\0';
 
-        /* ACPI _SUN — slot user number */
-        r = sd_device_new_from_subsystem_sysname(&pci, "subsystem", "pci");
-        if (r < 0)
-                return r;
-
-        r = sd_device_get_syspath(pci, &syspath);
-        if (r < 0)
-                return r;
-        if (!snprintf_ok(slots, sizeof slots, "%s/slots", syspath))
-                return -ENAMETOOLONG;
-
-        dir = opendir(slots);
-        if (!dir)
-                return -errno;
-
-        hotplug_slot_dev = names->pcidev;
-        while (hotplug_slot_dev) {
-                if (sd_device_get_sysname(hotplug_slot_dev, &sysname) < 0)
-                        continue;
-
-                FOREACH_DIRENT_ALL(dent, dir, break) {
-                        unsigned i;
-                        char str[PATH_MAX];
-                        _cleanup_free_ char *address = NULL;
-
-                        if (dot_or_dot_dot(dent->d_name))
-                                continue;
-
-                        r = safe_atou_full(dent->d_name, 10, &i);
-                        if (r < 0 || i <= 0)
-                                continue;
-
-                        /* match slot address with device by stripping the function */
-                        if (snprintf_ok(str, sizeof str, "%s/%s/address", slots, dent->d_name) &&
-                            read_one_line_file(str, &address) >= 0 &&
-                            startswith(sysname, address)) {
-                                hotplug_slot = i;
-                                break;
-                        }
-                }
-                if (hotplug_slot > 0)
-                        break;
-                if (sd_device_get_parent_with_subsystem_devtype(hotplug_slot_dev, "pci", NULL, &hotplug_slot_dev) < 0)
-                        break;
-                rewinddir(dir);
-        }
-
-        if (hotplug_slot > 0) {
+        /* figure out PCI hotplug index, but use it as part of the name only if
+           there is no other device claiming the same index */
+        hotplug_slot = pci_hotplug_slot_index(names->pcidev);
+        if (hotplug_slot > 0 && !pci_hotplug_slot_conflict(names->pcidev, hotplug_slot)) {
```

Comment:
Maybe we should propagate negative errno by `pci_hotplug_slot_index()` and `_conflict()`.